### PR TITLE
feat(gen2-migration): add dynamodb support for storage category

### PR DESCRIPTION
**Pull request by dgandhi62**
### Summary
Implemented full DynamoDB table migration support for Amplify Gen1 to Gen2 migration, enabling automatic conversion of DynamoDB resources from CloudFormation-based Gen1 projects to CDK-based Gen2 backend definitions.

### Key Features Added

**DynamoDB Table Discovery & Parsing:**
- Added `AppStorageDefinitionFetcher` to extract DynamoDB table configurations from Gen1 amplify-meta.json and cli-inputs.json
- Implemented AWS SDK integration to fetch live table schemas using `DescribeTableCommand`
- Support for parsing table attributes, key schemas, billing modes, and stream configurations

**Table Schema Migration:**
- **Partition & Sort Keys**: Automatic detection and conversion of primary key structures
- **Global Secondary Indexes (GSIs)**: Full GSI migration with partition/sort key mapping
- **Billing Modes**: Support for both `PAY_PER_REQUEST` and `PROVISIONED` with capacity units
- **DynamoDB Streams**: Stream configuration migration with view type preservation
- **Attribute Type Mapping**: Conversion from DynamoDB types (S, N, B) to CDK AttributeType enums

**Code Generation:**
- **CDK Table Creation**: Generates proper `new Table()` constructs with all configurations
- **Variable Name Sanitization**: Converts table names with hyphens to valid JavaScript identifiers
- **Import Management**: Automatic addition of required CDK imports (`Table`, `AttributeType`, `BillingMode`, `StreamViewType`)
- **Stack Integration**: Smart stack creation/reuse when S3 storage exists

**Comprehensive Test Coverage:**
- Unit tests for table parsing, GSI handling, Lambda permissions, and stream configurations
- Mock AWS SDK responses with proper TypeScript typing
- Edge case handling for multiple tables and mixed storage scenarios

### Technical Implementation

**Storage Definition Structure:**
```typescript
interface DynamoDBTableDefinition {
  tableName: string;
  partitionKey: { name: string; type: 'STRING' | 'NUMBER' | 'BINARY' };
  sortKey?: { name: string; type: 'STRING' | 'NUMBER' | 'BINARY' };
  gsis?: GlobalSecondaryIndex[];
  billingMode: 'PAY_PER_REQUEST' | 'PROVISIONED';
  readCapacity?: number;
  writeCapacity?: number;
  streamEnabled?: boolean;
  streamViewType?: string;
  lambdaPermissions?: LambdaPermission[];
  triggerFunctions?: string[];
}
```

Generated code example

```
const userTable_dev = new Table(storageStack, "userTable_dev", {
  partitionKey: { name: "userId", type: AttributeType.STRING },
  sortKey: { name: "timestamp", type: AttributeType.NUMBER },
  billingMode: BillingMode.PAY_PER_REQUEST,
  stream: StreamViewType.NEW_AND_OLD_IMAGES
});
```

The following still needs to be generated

**Lambda Integration:**
- **Environment Variables**: Automatic detection of Lambda functions accessing tables via env vars
- **IAM Permissions**: Generation of `grantReadWriteData` calls for Lambda table access
- **Stream Triggers**: Support for DynamoDB stream-triggered Lambda functions with proper permissions

```
userTable_dev.grantReadWriteData(backend.myFunction.resources.lambda);
backend.myFunction.addEnvironment("USER_TABLE_NAME", userTable_dev.tableName);
```

### CODE REVIEW

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.test.ts [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+      const result = synthesizer.render({
+        storage: {
+          importFrom: './storage/resource',
+          dynamoTables: [tableDefinition],
+        },
+      });
+
+      const source = printNodeArray(result);
+      expect(source).toContain('partitionKey: { name: "id", type: AttributeType.STRING }');
+      expect(source).toContain('sortKey: { name: "timestamp", type: AttributeType.NUMBER }');
```
Comment (iliapolo) [inline packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.test.ts]: Replace assertions in this file with `expect(source).toMatchInlineSnapshot()`. It will show us the entire output file and make it easier to review. 

See https://jestjs.io/docs/snapshot-testing

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts [hunk 2]
```diff
@@ -37,6 +38,7 @@ export interface BackendRenderParameters {
   storage?: {
     importFrom: string;
     dynamoDB?: string;
+    dynamoTables?: DynamoDBTableDefinition[];
```
Comment (iliapolo) [inline packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts:42]: See if we can remove the `dynamoDB` property - probably left over from the old codebase.

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/adapters/storage/gen1_storage_codegen_adapter.ts [hunk 3]
```diff
@@ -1,15 +1,53 @@
...
+  partitionKey: DynamoDBAttribute;
+  sortKey?: DynamoDBAttribute;
+};
+
+export type DynamoDBTableDefinition = {
+  tableName: string;
+  partitionKey: DynamoDBAttribute;
+  sortKey?: DynamoDBAttribute;
+  gsis?: DynamoDBGSI[];
+  lambdaPermissions?: {
```
Comment (iliapolo) [inline packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/adapters/storage/gen1_storage_codegen_adapter.ts]: Still not fully supported so please remove it

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/adapters/storage/gen1_storage_codegen_adapter.ts [hunk 4]
```diff
@@ -1,15 +1,53 @@
...
+    functionName: string;
+    envVarName: string;
+  }[];
+  triggerFunctions?: string[];
```
Comment (iliapolo) [inline packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/adapters/storage/gen1_storage_codegen_adapter.ts]: same

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts [hunk 5]
```diff
@@ -826,18 +834,269 @@ export class BackendSynthesizer {
...
-          ]),
+          ],
+          ts.NodeFlags.Const,
         ),
       );
+      nodes.push(stackDeclaration);
+
+      // Generate tables
+      renderArgs.storage.dynamoTables.forEach((table: DynamoDBTableDefinition) => {
+        const sanitizedTableName = this.sanitizeVariableName(table.tableName);
```
Comment (iliapolo) [inline packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts]: Remove the `-${env}` suffix from the table name before sanitizing. If we can avoid environment specifiers in the gen2 code we should. i.e instead of `tableName-main` we should just have `tableName`. 

It makes the gen2 code less specific and more portable since it doesn't "hint" at any specific environment names.

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts [hunk 6]
```diff
@@ -826,18 +834,269 @@ export class BackendSynthesizer {
...
+              ),
+              undefined,
+              [factory.createObjectLiteralExpression(gsiProps)],
+            ),
+          );
+          nodes.push(gsiCall);
+        });
+
+        // Add Lambda permissions
+        table.lambdaPermissions?.forEach((perm) => {
```
Comment (iliapolo) [inline packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts]: Not fully supported yet - remove

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts [hunk 7]
```diff
@@ -826,18 +834,269 @@ export class BackendSynthesizer {
...
+          const grantCall = factory.createExpressionStatement(
```
Comment (iliapolo) [inline packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts]: same

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts [hunk 8]
```diff
@@ -826,18 +834,269 @@ export class BackendSynthesizer {
...
+              ]),
+            ),
+          );
+        }
+
+        const tableDeclaration = factory.createVariableStatement(
+          [],
+          factory.createVariableDeclarationList(
+            [
+              factory.createVariableDeclaration(
```
Comment (iliapolo) [inline packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts]: Don't create a variable if its not needed further down the code. It creates eslint violations and/or compilation errors.

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/codegen-head/app_storage_definition_fetcher.ts [hunk 9]
```diff
@@ -58,12 +61,131 @@ export class AppStorageDefinitionFetcher {
     return triggers;
   };
 
+  // Check the properties
+  private parseDynamoDBTable = async (
```
Comment (iliapolo) [inline packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/codegen-head/app_storage_definition_fetcher.ts]: `parse` usually means operating on a string, not fetching values from an external source.  

```suggestion
  private fetchDynamoDBTable = async (
```

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/core/migration-pipeline.ts [hunk 10]
```diff
@@ -106,6 +107,26 @@ export interface Gen2RenderingOptions {
  */
 const createFileWriter = (path: string) => async (content: string) => fs.writeFile(path, content);
 
+/**
+ * Transforms handler file content from CommonJS to ES modules
+ * @param content - Original handler file content
+ * @returns Transformed content with ES module syntax
+ */
+const transformHandlerToESModule = (content: string): string => {
```
Comment (iliapolo) [inline packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/core/migration-pipeline.ts]: Seems unrelated.

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/core/migration-pipeline.ts [hunk 11]
```diff
@@ -56,6 +56,7 @@
   StorageTriggerEvent,
   ServerSideEncryptionConfiguration,
 } from '../generators/storage';
+import { DynamoDBTableDefinition } from '../adapters/storage';
```
[the remark this answers is not in the document]
Comment (iliapolo) [reply packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/core/migration-pipeline.ts]: Address this

**Review by iliapolo — CHANGES_REQUESTED** [review]
Please also update the README of app-2 now that storage is generated. Mention what step in the migration process will not work.

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts (hunk 2 above)
Comment (dgandhi62) [reply packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts:42]: Yes. It was being used to print an escape hatch that said "migrate tables with {names} manually". Removed it

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/core/migration-pipeline.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (dgandhi62) [reply packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/core/migration-pipeline.ts]: removed

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/core/migration-pipeline.ts (hunk 10 above)
Comment (dgandhi62) [reply packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/core/migration-pipeline.ts]: This was part of the functions implementation I had attempted. Removing this

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts (hunk 8 above)
Comment (dgandhi62) [reply packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts]: The variable is needed. For example - 
```
const MyTable = new Table(storageStack, 'MyTable', {
  partitionKey: { name: 'id', type: AttributeType.S },
  billingMode: BillingMode.PROVISIONED,
  readCapacity: 5,
  writeCapacity: 5,
  stream: StreamViewType.NEW_AND_OLD_IMAGES
});

// GSI operations
MyTable.addGlobalSecondaryIndex({
  indexName: 'GSI1',
  partitionKey: { name: 'gsi1pk', type: AttributeType.S },
  readCapacity: 5,
  writeCapacity: 5
});
```

Source - https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_dynamodb.Table.html

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts (hunk 5 above)
Comment (dgandhi62) [reply packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts]: Made the change

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.test.ts (hunk 1 above)
Comment (dgandhi62) [reply packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.test.ts]: Added

File: packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts (hunk 8 above)
Comment (iliapolo) [reply packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts]: But what if there are no GSIs to be added? this code should check if the variable is needed in the current generation context.

Comment (dgandhi62) [reply packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts]: I can conditionally create it, sure. Will update

Comment (dgandhi62) [reply packages/amplify-cli/src/commands/gen2-migration/codegen-generate/src/backend/synthesizer.ts]: Resolved

**Review by iliapolo — APPROVED** [review]
